### PR TITLE
Pyinvoke

### PR DESCRIPTION
**Pull request by mdaniel**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally prior to submission with `brew install <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [X] Does your submission pass `brew audit --new-formula <formula>` (after doing `brew install <formula>`)? (well, yes, modulo issue 5249)

---

This formula started life at version 0.12.2 because that was the specific version required to build another piece of software. However, 0.12 is not the latest version; that is 0.13.0 and is the one at the head of this PR. But I thought it would be useful to leave 0.12.2 in the history in case someone also wishes to build using the older version of pyinvoke

As you might suspect, I purposefully did not call the formula "invoke", however, I had mixed feelings about leaving their `inv` and `invoke` on the `$PATH`, so if you have guidance around that I'd welcome it.

### CODE REVIEW

File: Formula/pyinvoke.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  homepage "http://pyinvoke.org/"
+  url "https://github.com/pyinvoke/invoke/archive/0.13.0.tar.gz"
+  sha256 "4942aa901618f99afa70271d7b15fd018da0f174178fbca4048354c785e7736a"
+
+  head "https://github.com/pyinvoke/invoke.git"
+
+  depends_on :python if MacOS.version <= :snow_leopard
+
+  def install
+    virtualenv_install_with_resources
```
Comment (MikeMcQuaid) [inline Formula/pyinvoke.rb]: You're not using any `resource`s here? Does this have an external dependencies? CC @ilovezfs @tdsmith to check I haven't misunderstood this.

Comment (mdaniel) [inline Formula/pyinvoke.rb]: Correct, but compare that one line (which does everything one would wish for a `setup.py` based installation) versus:

```ruby
  def install
    ENV.prepend_create_path "PYTHONPATH", "#{libexec}/lib/python2.7/site-packages"

    system "python", "setup.py", "install", "--prefix=#{libexec}"
    bin.install Dir[libexec/"bin/*"]
    bin.env_script_all_files(libexec+"bin", :PYTHONPATH => ENV["PYTHONPATH"])
  end
```

It's not me being lazy; I can copy-and-paste with the best of them, but I think it's a vast legibility improvement. I was unable to find anything in [the Formula Cookbook](https://github.com/Homebrew/brew/blob/master/docs/Formula-Cookbook.md#specifying-gems-python-modules-go-projects-etc-as-dependencies) that implied one should favor the more verbose approach, so I didn't.

Comment (MikeMcQuaid) [inline Formula/pyinvoke.rb]: Yeh, not really your issue here just interested what @tdsmith thinks re: naming etc.

**Comment by tdsmith** [comment]
It looks okay to me. It's installing all zero of the resources :)

**Comment by MikeMcQuaid** [comment]
@tdsmith Last question: is it worth bottling things like this or better just have them `bottle :unneeded`?

**Comment by tdsmith** [comment]
Doesn't hurt but pure-Python modules with no dependencies don't gain any appreciable benefits from bottling.

File: Formula/pyinvoke.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+      @task
+      def clean(ctx, extra=''):
+          patterns = ['foo']
+          if extra:
+              patterns.append(extra)
+          for pattern in patterns:
+              run("rm -rf {}".format(pattern))
+    EOS
+    mkdir_p testpath/"foo"/"bar"
```
Comment (DomT4) [inline Formula/pyinvoke.rb]: `(testpath/"foo/bar").mkpath` is preferred.

File: Formula/pyinvoke.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    mkdir_p testpath/"baz"
```
Comment (DomT4) [inline Formula/pyinvoke.rb]: And `(testpath/"baz").mkpath`

File: Formula/pyinvoke.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    system bin/"invoke", "clean"
+    assert !File.exist?(testpath/"foo"), "\"pyinvoke clean\" should have deleted \"foo\""
```
Comment (DomT4) [inline Formula/pyinvoke.rb]: Doesn't need to be decided right here, but now we're enforcing Ruby 2.0 we can use things like `assert_predicate` if we like, which produces nicer error messages by default without necessarily having to explicitly write out one.

File: Formula/pyinvoke.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
+class Pyinvoke < Formula
+  include Language::Python::Virtualenv
+
+  desc "Pythonic task management & command execution"
+  homepage "http://pyinvoke.org/"
+  url "https://github.com/pyinvoke/invoke/archive/0.13.0.tar.gz"
+  sha256 "4942aa901618f99afa70271d7b15fd018da0f174178fbca4048354c785e7736a"
+
```
Comment (DomT4) [inline Formula/pyinvoke.rb]: No empty line required here.

**Review by DomT4 — CHANGES_REQUESTED** [review]

File: Formula/pyinvoke.rb (hunk 4 above)
Comment (mdaniel) [inline Formula/pyinvoke.rb]: How does one `assert_predicate thing, *negated* operation`? I tried the simplistic `a_p thing, !:exist?` and got a horrible explosion and I didn't see any `:missing?` on `File`

I only found [one example](https://github.com/Homebrew/homebrew-core/blob/0379d0d/Formula/conan.rb#L141) across all the `homebrew-*` Taps I have locally, and [the documentation](http://ruby-doc.org/stdlib-2.0.0/libdoc/minitest/rdoc/MiniTest/Assertions.html#assert_predicate-method) wasn't helpful.

Comment (DomT4) [inline Formula/pyinvoke.rb]: That would be `refute_predicate`. Probably worth me polling @ilovezfs and @zmwangx on this change though, I'm not around as much as I used to be so they'll end up dealing with it day to day if we decide to change.

**Comment by MikeMcQuaid** [comment]
@BrewTestBot test this please

**Comment by MikeMcQuaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

For future reference the preferred commit message format for simple version updates is `pyinvoke 0.13.0 (new formula)`. Please complete the issue template in future PRs where this format is detailed in the linked [Contributing guidelines](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md). Thanks!

**Pull request closed without merging**